Overhaul logs, and move to /var/log/archivematica

mistydemeo opened the pull request.

Commit:
Move Archivematica logs to /var/log/archivematica

refs #6647

mistydemeo commented:
Note that this does _not_ currently migrate the MCP server and client logs to use Python's logging module. That would unfortunately be a pretty major task, since it would require replacing all print statements in those two components. We can work slowly towards that.

This creates a new `/var/log/archivematica` directory, which is owned by `archivematica:archivematica`. It's group-writeable and setgid, so any user of the Archivematica group should be able to create and edit logs without causing problems for other users.

Hwesta commented on localDevSetup/createLocalDevDirectories.sh:
```diff
@@ -32,6 +32,10 @@ etc="/etc/archivematica"
+log="/var/log/archivematica"
+sudo mkdir $log
+sudo chown -R archivematica:archivematica $log
+sudo chmod g+ws $log
```
Ask Justin about making sure this happens for the MCPClient package install as well.

Commit:
Logging: adjust namespaces for logger objects

refs #6647

Commit:
Add custom log handler

This allows certain files which may be called as multiple users to
safely log to the same file, such as client scripts.

refs #6647

mistydemeo commented:
Updated to use a custom logging handler which correctly sets group-writeability of the logs.

We may still want to update to prevent certain files from hardcoding their logging paths. e.g. things that are imported and not executed directly should perhaps just `getLogger` without setting a logger.

mistydemeo commented:
Rebased onto qa/1.x.

Hwesta commented on src/MCPClient/lib/clientScripts/getAipStorageLocations.py:
```diff
@@ -1,15 +1,24 @@
+path = '/usr/share/archivematica/dashboard'
+if path not in sys.path:
+    sys.path.append(path)
+os.environ['DJANGO_SETTINGS_MODULE'] = 'settings.common'
```
This should be rebased, since this isn't needed anymore.

mistydemeo replied:
Oops, thanks - it was reintroduced by the rebase, I'll remove.

Hwesta commented on src/MCPClient/lib/clientScripts/moveToBacklog.py:
```diff
@@ -6,11 +6,13 @@
-logger = logging.getLogger(__name__)
-logging.basicConfig(filename="/tmp/archivematica.log",
-    level=logging.INFO)
+logger = logging.getLogger('archivematica.dashboard')
```
Shouldn't this be `'archivematica.mcp.client'`?

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -45,9 +46,10 @@
-logger = logging.getLogger(__name__)
-logging.basicConfig(filename="/tmp/archivematicaDashboard.log",
-    level=logging.INFO)
+logger = logging.getLogger('archivematica.dashboard')
```
archivematica.common?

Hwesta commented on src/archivematicaCommon/lib/externals/requests_1_20/packages/urllib3/__init__.py:
```diff
@@ -46,7 +46,7 @@ def add_stderr_logger(level=logging.DEBUG):
     # This method needs to be in this __init__.py to get the __name__ correct
     # even if urllib3 is vendored within another package.
-    logger = logging.getLogger(__name__)
+    logger = logging.getLogger('archivematica.common')
```
Do we need to update externals logging?

mistydemeo replied:
Probably not, I'll change this back.

Hwesta commented on src/archivematicaCommon/lib/storageService.py:
```diff
@@ -8,9 +8,13 @@
-logger = logging.getLogger(__name__)
-logging.basicConfig(filename="/tmp/archivematica.log",
-    level=logging.INFO)
+sys.path.append("/usr/lib/archivematica/archivematicaCommon")
```
Isn't this already in the path?

mistydemeo replied:
This is another case where the rebase reintroduced some deleted code, sorry - removing!

mistydemeo commented:
On Justin's suggestion, updated the log paths; they now use subdirectories. Logs are written to:

```
/var/log/archivematica/archivematica.log
/var/log/archivematica/dashboard/dashboard.log
/var/log/archivematica/MCPClient/{hostname}.log
/var/log/archivematica/MCPServer/{hostname-date}.log
```

The first two are rotated, while the Client and Server logs are not.

mistydemeo commented:
Updated.

Hwesta commented on src/MCPClient/lib/clientScripts/moveToBacklog.py:
```diff
@@ -6,11 +6,13 @@
-logging.basicConfig(filename="/tmp/archivematica.log",
-    level=logging.INFO)
+logger = logging.getLogger('archivematica.mcp.client')
+logger.addHandler(GroupWriteRotatingFileHandler("/var/log/archivematica/dashboard/dashboard.log",
```
Log to archivematica.log? Maybe `/var/log/archivematica/MCPClient/???.log`?

Also true for all the other MCPClient logs.

mistydemeo replied:
Yeah, maybe these should log to `/var/log/archivematica/MCPClient/{scriptname}.log`?

Hwesta replied:
I think I'd rather `/var/log/archivematica/MCPClient/clientscripts.log` and have the logger be based on the filename.

Well, really I'd rather have this configured once in the `__init__.py` and everything just does `LOGGER = logging.getLogger(__file__)`, but baby steps.

Hwesta commented on src/MCPClient/lib/clientScripts/upload-archivistsToolkit-fcd.py:
```diff
@@ -35,9 +35,9 @@
+logger = logging.getLogger('archivematica.mcp.client')
 logger.addHandler(logging.NullHandler())
-logger.addHandler(logging.FileHandler('/tmp/at_upload.log', mode='a'))
+logger.addHandler(logging.FileHandler('/var/log/archivematica/at_upload.log', mode='a'))
```
Should these be `GroupWriteRotatingFileHandler`?  Or do they not get run enough to care?

Hwesta commented on src/archivematicaCommon/lib/archivistsToolkit/atk.py:
```diff
@@ -2,7 +2,7 @@
 logging.basicConfig(level=logging.INFO)
-logger = logging.getLogger(__name__)
+logger = logging.getLogger('archivematica.common')
 logger.addHandler(logging.NullHandler())
```
Only a NullHandler for this?

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -45,9 +46,10 @@
-logging.basicConfig(filename="/tmp/archivematicaDashboard.log",
-    level=logging.INFO)
+logger = logging.getLogger('archivematica.common')
+logger.addHandler(GroupWriteRotatingFileHandler("/var/log/archivematica/dashboard/dashboard.log",
```
Log to archivematica.log? Or don't set a handler and let the importer decide? Same for all the archivematicaCommon loggers.

mistydemeo replied:
Yeah, I think that makes the most sense.

mistydemeo replied:
Yeah, letting the importer decide works better I think.

Hwesta commented on src/archivematicaCommon/lib/utilities/restart-elasticsearch-if-down:
```diff
@@ -25,7 +25,7 @@ except pyes.exceptions.NoServerAvailable:
 
     # note service was restarted in log
-    log_file = "/tmp/archivematica_elasticsearch_restarter.log"
+    log_file = "/var/log/archivematica/elasticsearch_restarter.log"
```
I though this was in archivematica-devtools?

mistydemeo replied:
Hm, I thought so too. Looks like it didn't get deleted from qa/1.x. I'll nuke it there.

mistydemeo replied:
artefactual/archivematica#181

Commit:
Delegate setting logger to importer

refs #6647

Commit:
Client scripts: add logging

refs #6647

mistydemeo commented on src/MCPClient/lib/clientScripts/getAipStorageLocations.py:
```diff
@@ -22,6 +19,8 @@ def get_aip_storage_locations(purpose):
 
 
 if __name__ == '__main__':
+    logger = get_script_logger("archivematica.mcp.client.getAipStorageLocations")
```
This is how I'm looking at handling the logs for client scripts. This function sets up a root logger for `archivematica` as well as a logger object for the specific script, and returns the latter. All of the scripts log to the same `client_scripts` log file, with the filename being at the beginning of the message to make it easier to tell where logging messages are coming from. An example:

> getAipStorageLocations.py: INFO     archivematica.common.get_location: Storage locations returned: [{'uuid': '516b19e7-cfd9-4c44-8f99-b1fe27bf390e', 'pipeline': ['/api/v2/pipeline/7600ac20-4f62-4de6-943d-7a2e8a29aa5a/', '/api/v2/pipeline/f61d9185-cd18-4d91-9c91-8087fbfe88bd/', '/api/v2/pipeline/294e6779-56f4-4468-8d9b-94b311d12eac/', '/api/v2/pipeline/be01d080-139c-4eec-9bfe-d037c70e160a/'], 'used': '4887384', 'description': 'Store AIP in standard Archivematica Directory', 'space': '/api/v2/space/3dcb3b17-790d-44e3-b948-943a62c6a80c/', 'enabled': True, 'quota': None, 'relative_path': 'var/archivematica/sharedDirectory/www/AIPsStore', 'purpose': 'AS', 'path': '/var/archivematica/sharedDirectory/www/AIPsStore', 'resource_uri': '/api/v2/location/516b19e7-cfd9-4c44-8f99-b1fe27bf390e/'}]

When we add logging within the client scripts themselves, I was thinking we'd pass a `logger` keyword argument into the `main` function (with the default being a logger with a null handler), rather than relying on a file-wide logger object like in library code.

Right now the log level is hardcoded, but we could look at making it configured in the MCPClient settings file and have the scripts fetch the level to pass into this function.

mistydemeo commented:
Adding logging to all client scripts.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS.py:
```diff
@@ -174,6 +175,8 @@ def createFileSec(path, file_group_identifier, base_path, base_path_name, parent
                 createFileSec(os.path.join(path, item), file_group_identifier, base_path, base_path_name, parentBranch, div, sip_uuid)
 
 if __name__ == '__main__':
+    logger = get_script_logger("archivematica.mcp.client.createMETS")
```
createTransferMETS?

mistydemeo replied:
I'd like to keep a correlation between the script name and logger name, but let's look at renaming the scripts themselves in a PR after this is merged!

Hwesta replied:
Reasonable

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -857,6 +858,8 @@ def create_object_metadata(struct_map, baseDirectoryPath):
 
 
 if __name__ == '__main__':
+    logger = get_script_logger("archivematica.mcp.client.createMETS2")
```
createSIPMETS / createAIPMETS?

Hwesta commented on src/archivematicaCommon/lib/custom_handlers.py:
```diff
@@ -0,0 +1,35 @@
+        finally:
+            os.umask(prevumask)
+
+STANDARD_FORMAT = "%(levelname)-8s  %(asctime)s  %(name)s.%(funcName)s:  %(message)s"
```
%(lineno)d for this one too?

Hwesta commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -41,12 +41,14 @@
+logger = logging.getLogger('archivematica.dashboard')
+logger.addHandler(GroupWriteRotatingFileHandler("/var/log/archivematica/dashboard/dashboard.log",
+     maxBytes=4194304))
+logger.setLevel(logging.INFO)
```
Configure all these in dashboard/src/settings/common.py instead?

mistydemeo commented:
Updated with changes based on feedback - going to merge if there aren't any more comments!

Commit:
Dashboard: configure logs in settings/common.py

refs #6647

Hwesta commented:
:shipit:
